src/core/thread/link_metrics_tlvs.hpp

```diff
@@ -0,0 +1,378 @@
+        {
+            SetLength(sizeof(*this) - sizeof(Tlv) - sizeof(uint32_t) + sizeof(uint8_t)); // The value is 1 byte long
+        }
+    }
+
+    /**
+     * This method returns the metric value in 8 bits.
+     *
+     * @returns The metric value.
+     *
+     */
+    uint8_t GetMetricsValue8(void) const { return mMetricsValue.m8; }
+
+    /**
+     * This method returns the metric value in 32 bits.
+     *
+     * @returns The metric value.
+     *
+     */
+    uint32_t GetMetricsValue32(void) const { return mMetricsValue.m32; }
+
+    /**
+     * This method sets the metric value(8 bits).
+     *
+     * @param[in]  aMetricsValue  Metrics value.
+     *
+     */
+    void SetMetricsValue8(uint8_t aMetricsValue) { mMetricsValue.m8 = aMetricsValue; }
+
+    /**
+     * This method sets the metric value(32 bits).
+     *
+     * @param[in]  aMetricsValue  Metrics value.
+     *
+     */
+    void SetMetricsValue32(uint32_t aMetricsValue) { mMetricsValue.m32 = aMetricsValue; }
+
+private:
+    LinkMetricsTypeId mMetricsTypeId;
+    union
+    {
+        uint8_t  m8;
+        uint32_t m32;
+    } mMetricsValue;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Link Metrics Query Id Sub-TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class LinkMetricsQueryId : public Tlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void)
+    {
+        SetType(kLinkMetricsQueryId);
+        SetLength(sizeof(*this) - sizeof(Tlv));
+    }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval TRUE   If the TLV appears to be well-formed.
+     * @retval FALSE  If the TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(Tlv); }
+
+    /**
+     * This method returns the Series ID.
+     *
+     * @returns The Series ID.
+     *
+     */
+    uint8_t GetSeriesId(void) const { return mSeriesId; }
+
+    /**
+     * This method sets the Series ID.
+     *
+     * @param[in]  aSeriesId  The Series ID.
+     *
+     */
+    void SetSeriesId(uint8_t aSeriesId) { mSeriesId = aSeriesId; }
+
+private:
+    uint8_t mSeriesId;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This class implements Link Metrics Query Options Sub-TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class LinkMetricsQueryOptions : public Tlv
+{
+public:
+    /**
+     * This method initializes the TLV.
+     *
+     */
+    void Init(void)
+    {
+        SetType(kLinkMetricsQueryOptions);
+        SetLength(sizeof(*this) - sizeof(Tlv) - sizeof(mMetricsTypeIds));
+    }
+
+    /**
+     * This method indicates whether or not the TLV appears to be well-formed.
+     *
+     * @retval TRUE   If the TLV appears to be well-formed.
+     * @retval FALSE  If the TLV does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const { return GetLength() <= sizeof(*this) - sizeof(Tlv); }
```

Comment:
For `kLinkMetricsQueryOptions` tlv, if included, should there be at least 1 Link Metrics Type ID Flags be included?

Reply:
For these comments to check values, I think we should handle it in `MLE` logic since the tlvs classes are for parsing. Thoughts?

Reply:
If it is agreed that at least 1 Link Metrics Type ID Flags should be included, I would take it as a part of length check, GetLength() should be >=1

Reply:
Updated the check.